core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java

```diff
@@ -530,7 +530,7 @@ public static RelDataType createJoinType(
         || (fieldNameList.size()
         == (systemFieldList.size()
         + leftType.getFieldCount()
-        + rightType.getFieldCount()));
+        + (rightType == null ? 0 : rightType.getFieldCount())));
```

Comment:
we require that rightType is not null. add precondition, rather than making code more defensive.